content/advent-2019/control-packetflow-tcp-nodelay.md

```diff
@@ -0,0 +1,226 @@
+
+In Go, `TCP_NODELAY` is enabled by default, but the standard library offers the ability to disable the behavior via the [net.SetNoDelay](https://golang.org/pkg/net/#TCPConn.SetNoDelay) method.
+
+## A small experiment
+
+To observe what's happening at the packet-level, and see the differences in packet arrival, we will use a tiny TCP client/server written in Go. Usually, we have inter-connected services across different regions, but for the sake of the experiment, we will experiment in our local machine. The full source code is also available on [Github](https://github.com/dastergon/gopheracademy-advent2019-tcp-no-delay).
+
+The server code (server.go):
+
+```go
+package main
+
+import (
+    "bufio"
+    "fmt"
+    "log"
+    "net"
+    "strings"
+)
+
+func main() {
+    port := ":" + "8000"
+
+    // Create a listening socket.
+    l, err := net.Listen("tcp", port)
+    if err != nil {
+        log.Fatal(err)
+    }
+    defer l.Close()
+
+    for {
+        // Accept new connections.
+        c, err := l.Accept()
+        if err != nil {
+            log.Println(err)
+            return
+        }
+
+        // Process newly accepted connection.
+        go handleConnection(c)
+    }
+}
+func handleConnection(c net.Conn) {
+    fmt.Printf("Serving %s\n", c.RemoteAddr().String())
+
+    for {
+        // Read what has been sent from the client.
+        netData, err := bufio.NewReader(c).ReadString('\n')
+        if err != nil {
+            log.Println(err)
+            return
+        }
+
+        cdata := strings.TrimSpace(netData)
+        if cdata == "GOPHER" {
+            c.Write([]byte("GopherAcademy Advent 2019!"))
+        }
+
+        if cdata == "EXIT" {
+            break
+        }
+    }
+    c.Close()
+}
+```
+
+The client code (client.go):
+
+```go
+package main
+
+import (
+    "fmt"
+    "log"
+    "net"
+)
+
+func main() {
+    target := "localhost:8000"
+
+    raddr, err := net.ResolveTCPAddr("tcp", target)
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    // Establish a connection with the server.
+    conn, err := net.DialTCP("tcp", nil, raddr)
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    fmt.Println("Sending Gophers down the pipe...")
+
+    for i := 1; i <= 5; i++ {
+        // Send the word "GOPHER" to the open connection.
+        _, err = conn.Write([]byte("GOPHER\n"))
+        if err != nil {
+            log.Fatal(err)
+        }
+    }
+}
+```
+
+To observe the behavior change, first execute `tcpdump`. You might have to
+change your network interface to match your own machine:
+
+    sudo tcpdump -X  -i lo0 'port 8000'
+
+Then, execute the server (server.go) and the client (client.go).
+
+    go run server.go
+
+In another terminal window execute:
+
+    go run client.go
+
+Initially, if we look closer at the payload, we'll notice that the word `"GOPHER"` is sent in 5 separate packets. For brevity, I just posted only a couple of packets.
```

Comment:
This sounded to me like it was splitting up the single word "GOPHER" into five packets somehow, as opposed to sending *each* instance of the word in a separate packet. Maybe "each write is transmitted as a separate packet" or something?

Reply:
You're right, updated!